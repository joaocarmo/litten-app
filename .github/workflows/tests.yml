# This workflow will do a clean install of node dependencies, build the source
# code and run the available tests
name: Tests

on:
  push:
    branches: [ main, next ]
  pull_request:
    branches: [ main, next ]

env:
  CI: true
  FIREBASE_EMULATORS_PATH: $GITHUB_WORKSPACE/emulator-cache

jobs:
  test:
    environment: development
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v3
    - name: Use Node.js v16 (LTS)
      uses: actions/setup-node@v3
      with:
        node-version: '16'
        cache: 'yarn'
    - name: Install firebase-tools globally
      run: yarn global add firebase-tools
    - name: Create .env file
      run: cp .env.example .env
    - name: Install dependencies
      run: yarn --frozen-lockfile
    - name: Lint the source files
      run: yarn lint
    - name: Type check
      run: yarn type-check
    - name: Validate the translations
      run: yarn translations-check
    - name: Execute the tests
      run: yarn test
    - name: Execute the firebase rules tests
      run: yarn test:firestore:rules
    - name: Execute the firebase model tests
      run: yarn test:firestore:model
    - name: Cache firebase emulators
      uses: actions/cache@v3
      with:
        path: ${{ env.FIREBASE_EMULATORS_PATH }}
        key: ${{ runner.os }}-firebase-emulators-${{ hashFiles('emulator-cache/*.{jar,zip}') }}
      continue-on-error: true
