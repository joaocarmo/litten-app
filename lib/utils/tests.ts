import { apps, clearFirestoreData } from '@firebase/rules-unit-testing'

export const projectId = 'litten-app'

export const cleanFirestore = async (): Promise<void> => {
  await Promise.all(apps().map((app) => app.delete()))
  await clearFirestoreData({
    projectId,
  })
}

export const mapName =
  <T>(key: string) =>
  (object: T): T =>
    Object.assign(object, {
      toString: function () {
        return this[key]
      },
    })

export const getRandInt = (min: num, max: num): num =>
  Math.floor(Math.random() * (max - min + 1) + min)
