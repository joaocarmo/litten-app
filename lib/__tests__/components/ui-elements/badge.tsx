import { Text } from 'react-native'
import { render } from '@utils/tests/react-native'
import { UIBadge, UIText } from '@ui-elements'

describe('Snapshot test for the "UIBadge" component', () => {
  const content = 'Text'

  it('renders correctly a positive number', () => {
    const badgeContent = 10
    const element = render(
      <UIBadge number={badgeContent}>
        <UIText>{content}</UIText>
      </UIBadge>,
    )
    const elementTextComponents = element.container.findAllByType(Text)
    const [elementBadgeContent, elementContent] = elementTextComponents

    expect(element.toJSON()).toMatchSnapshot()
    expect(elementContent.props.children).toBe(content)
    expect(elementBadgeContent.props.children).toBe(badgeContent)
  })

  it('renders correctly without zero', () => {
    const badgeContent = 0
    const element = render(
      <UIBadge number={badgeContent}>
        <UIText>{content}</UIText>
      </UIBadge>,
    )
    const elementTextComponents = element.container.findAllByType(Text)
    const [elementContent] = elementTextComponents

    expect(element.toJSON()).toMatchSnapshot()
    expect(elementContent.props.children).toBe(content)
  })

  it('renders correctly without a negative number', () => {
    const badgeContent = -10
    const element = render(
      <UIBadge number={badgeContent}>
        <UIText>{content}</UIText>
      </UIBadge>,
    )
    const elementTextComponents = element.container.findAllByType(Text)
    const [elementContent] = elementTextComponents

    expect(element.toJSON()).toMatchSnapshot()
    expect(elementContent.props.children).toBe(content)
  })

  it('renders correctly with zero', () => {
    const badgeContent = 0
    const element = render(
      <UIBadge number={badgeContent} showNonPositive>
        <UIText>{content}</UIText>
      </UIBadge>,
    )
    const elementTextComponents = element.container.findAllByType(Text)
    const [elementBadgeContent, elementContent] = elementTextComponents

    expect(element.toJSON()).toMatchSnapshot()
    expect(elementContent.props.children).toBe(content)
    expect(elementBadgeContent.props.children).toBe(badgeContent)
  })

  it('renders correctly with a negative number', () => {
    const badgeContent = -10
    const element = render(
      <UIBadge number={badgeContent} showNonPositive>
        <UIText>{content}</UIText>
      </UIBadge>,
    )
    const elementTextComponents = element.container.findAllByType(Text)
    const [elementBadgeContent, elementContent] = elementTextComponents

    expect(element.toJSON()).toMatchSnapshot()
    expect(elementContent.props.children).toBe(content)
    expect(elementBadgeContent.props.children).toBe(badgeContent)
  })
})
