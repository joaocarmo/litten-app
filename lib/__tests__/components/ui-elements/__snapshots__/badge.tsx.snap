// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test for the "UIBadge" component renders correctly a positive number 1`] = `
<View
  style={
    {
      "position": "relative",
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#e30000",
        "borderRadius": 20,
        "justifyContent": "center",
        "minHeight": 20,
        "minWidth": 20,
        "padding": 5,
        "position": "absolute",
        "right": -15,
        "top": -7.5,
        "zIndex": 1,
      }
    }
  >
    <Text
      style={
        {
          "color": "#ffffff",
          "fontSize": 9.832647462277093,
          "fontWeight": "700",
          "textAlign": "center",
        }
      }
    >
      10
    </Text>
  </View>
  <Text
    style={
      [
        {
          "color": "#000000",
          "fontSize": 14,
          "fontWeight": "200",
        },
        {
          "marginBottom": 8,
          "marginTop": 8,
          "paddingBottom": 6,
          "paddingTop": 6,
        },
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    Text
  </Text>
</View>
`;

exports[`Snapshot test for the "UIBadge" component renders correctly with a negative number 1`] = `
<View
  style={
    {
      "position": "relative",
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#e30000",
        "borderRadius": 20,
        "justifyContent": "center",
        "minHeight": 20,
        "minWidth": 20,
        "padding": 5,
        "position": "absolute",
        "right": -15,
        "top": -7.5,
        "zIndex": 1,
      }
    }
  >
    <Text
      style={
        {
          "color": "#ffffff",
          "fontSize": 9.832647462277093,
          "fontWeight": "700",
          "textAlign": "center",
        }
      }
    >
      -10
    </Text>
  </View>
  <Text
    style={
      [
        {
          "color": "#000000",
          "fontSize": 14,
          "fontWeight": "200",
        },
        {
          "marginBottom": 8,
          "marginTop": 8,
          "paddingBottom": 6,
          "paddingTop": 6,
        },
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    Text
  </Text>
</View>
`;

exports[`Snapshot test for the "UIBadge" component renders correctly with zero 1`] = `
<View
  style={
    {
      "position": "relative",
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#e30000",
        "borderRadius": 20,
        "justifyContent": "center",
        "minHeight": 20,
        "minWidth": 20,
        "padding": 5,
        "position": "absolute",
        "right": -15,
        "top": -7.5,
        "zIndex": 1,
      }
    }
  >
    <Text
      style={
        {
          "color": "#ffffff",
          "fontSize": 9.832647462277093,
          "fontWeight": "700",
          "textAlign": "center",
        }
      }
    >
      0
    </Text>
  </View>
  <Text
    style={
      [
        {
          "color": "#000000",
          "fontSize": 14,
          "fontWeight": "200",
        },
        {
          "marginBottom": 8,
          "marginTop": 8,
          "paddingBottom": 6,
          "paddingTop": 6,
        },
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    Text
  </Text>
</View>
`;

exports[`Snapshot test for the "UIBadge" component renders correctly without a negative number 1`] = `
<View
  style={
    {
      "position": "relative",
    }
  }
>
  <Text
    style={
      [
        {
          "color": "#000000",
          "fontSize": 14,
          "fontWeight": "200",
        },
        {
          "marginBottom": 8,
          "marginTop": 8,
          "paddingBottom": 6,
          "paddingTop": 6,
        },
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    Text
  </Text>
</View>
`;

exports[`Snapshot test for the "UIBadge" component renders correctly without zero 1`] = `
<View
  style={
    {
      "position": "relative",
    }
  }
>
  <Text
    style={
      [
        {
          "color": "#000000",
          "fontSize": 14,
          "fontWeight": "200",
        },
        {
          "marginBottom": 8,
          "marginTop": 8,
          "paddingBottom": 6,
          "paddingTop": 6,
        },
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    Text
  </Text>
</View>
`;
